src/client/coll/mod.rs

```diff
@@ -77,30 +79,15 @@ impl<'a> Collection<'a> {
     }
 
     /// Returns a list of documents within the collection that match the filter.
-    pub fn find(&self, filter: Option<bson::Document>, options: Option<FindOptions>)
-                -> Result<Vec<bson::Document>, String> {
+    pub fn find(&'a self, filter: Option<bson::Document>, options: Option<FindOptions>)
+                -> Result<Cursor<'a>, String> {
 
         let doc = filter.unwrap_or(bson::Document::new());
         let options = options.unwrap_or(FindOptions::new());
         let flags = OpQueryFlags::with_find_options(&options);
 
-        let req = try!(Message::with_query(self.get_req_id(), flags, self.namespace.to_owned(),
-                                           options.skip as i32, options.limit, doc, options.projection));
-
-        let socket = match self.db.client.socket.lock() {
-            Ok(val) => val,
-            _ => return Err("Client socket lock poisoned.".to_owned()),
-        };
-
-        try!(req.write(&mut *socket.borrow_mut()));
-        let message = try!(Message::read(&mut *socket.borrow_mut()));
-
-        match message {
-            Message::OpReply { header: _, flags: _, cursor_id: _,
-                               starting_from: _, number_returned: _,
-                               documents, } => Ok(documents),
-            _ => Err("Invalid response received from server".to_owned())
-        }
+        Cursor::query_with_batch_size(self, options.batch_size, flags, options.skip as i32,
```

Comment:
I think we need to special case `options.limit = 1` somewhere, because with [OpQuery, when `number_to_return` is 1, the server assumes its a find_one and kills the cursor](http://docs.mongodb.org/meta-driver/latest/legacy/mongodb-wire-protocol/#op-query)